tslearn/early_classification.py

```diff
@@ -0,0 +1,266 @@
+            (vector_of_ones, (y_, c_k)),
+            shape=(self.__n_classes_, self.n_clusters),
+        ).toarray()
+        self.pyck_ /= self.pyck_.sum(axis=0, keepdims=True)
+
+        for t in range(self.minimum_time_stamp, self.__len_X_ + 1):
+            self.classifiers_[t].fit(X1[:, :t], y1)
+            for k in range(0, self.n_clusters):
+                index = (c_k2 == k)
+                if index.shape[0] != 0:
+                    X2_current_cluster = X2[index, :t]
+                    y2_current_cluster = y2[index]
+                    y2_hat = self.classifiers_[t].predict(
+                        X2_current_cluster[:, :t]
+                    )
+                    conf_matrix = confusion_matrix(
+                        y2_current_cluster, y2_hat, labels=y_classes_indices, normalize="true"
+                    )
+                    #The diagonal of the confusion matrices must be put to 0
+                    # to fullfil the criterions of the cost function
+                    np.fill_diagonal(conf_matrix, 0)
+                    self.pyhatyck_[t - self.minimum_time_stamp, k] = conf_matrix
+
+    def get_cluster_probas(self, X):
+        """
+        This function computes the probabilities of the time series xt to be in a cluster of the model
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated
+
+        Returns
+        -------
+        vector : the probabilities of the given time series to be in the clusters of the model
+        """
+        diffs = X[np.newaxis, :] - self.cluster_.cluster_centers_[:, :len(X)]
+        distances_clusters = np.linalg.norm(diffs, axis=(1, 2))
+        minimum_distance = np.min(distances_clusters)
+        average_distance = np.mean(distances_clusters)
+        minimum_distance_to_average = (
+            average_distance - minimum_distance
+        ) / average_distance
+        delta = (average_distance - distances_clusters) / average_distance
+        probas = 1.0 / (
+            np.exp(self.lamb * minimum_distance_to_average)
+            + np.exp(self.lamb * (minimum_distance_to_average - delta))
+            + 1e-6
+        )
+        return probas / probas.sum()
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[-1]
+        sum_pyhatyck = np.sum(self.pyhatyck_[truncated_t + tau - self.minimum_time_stamp], axis=1)
+        sum_pyhatyck = np.transpose(sum_pyhatyck)
```

Comment:
Same remark here: if you need to transpose, it means that the ordering is not the same as in `self.pyck_` which is not natural, so it should be documented.

Reply:
I think the same answer applies here, for clarity purposes in the building of self.pyhatyck the dimensions are ordered in a way that forces here the transposition to match with those of self.pyck_. However I will investigate whether self.pyck_ could be built another way to prevent the transposition.